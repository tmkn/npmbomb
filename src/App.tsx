/** @jsx jsx */
import { jsx, css, Global, SerializedStyles } from "@emotion/react";
import React, { useState, lazy, Suspense } from "react";
import { BrowserRouter, Redirect, Route, Switch } from "react-router-dom";

const Index = lazy(() => import("./components/index/Index"));
const Package = lazy(() => import("./components/package/Package"));
const Results = lazy(() => import("./components/results/Results"));
const Overview = lazy(() => import("./components/overview/Overview"));

import { AppContext, IAppContext, IAppState, IGuessResult } from "./AppContext";
import Header from "./components/shared/header/Header";
import Content from "./components/shared/content/Content";
import Footer from "./components/shared/footer/Footer";
import { LoadingIndicator } from "./components/shared/loading/LoadingIndicator";
import { serifFont, sansSerifFont, primaryColor, secondaryColor, globalFocusStyle } from "./css";
import IconFont from "./codicon.ttf";
import "./codicon.css";

const globalStyle: SerializedStyles = css`
    @import url("https://fonts.googleapis.com/css2?family=Roboto+Mono:wght@300&display=swap");

    body {
        background: white;
        margin: 0;
        padding: 0;
        min-height: 100vh;
        max-width: 100vw;
        display: flex;
        flex-direction: column;
        font-family: "${sansSerifFont}";
    }

    a {
        text-decoration: none;
    }

    h1 {
        font-family: "${serifFont}";
        color: ${primaryColor};
        font-weight: 200;
    }

    h2 {
        font-family: "${serifFont}";
        color: ${secondaryColor};
        font-weight: 200;
    }

    @font-face {
        font-family: "IconFont";
        src: url(${IconFont}) format("ttf");
    }

    @font-face {
        font-family: "Muli";
        font-style: normal;
        font-weight: 700;
        font-display: swap;
        src: url(https://fonts.gstatic.com/s/muli/v17/7Aulp_0qiz-aVz7u3PJLcUMYOFkpl0k30e6fwniDtzM.woff)
            format("woff");
        unicode-range: U+0000-00FF, U+0131, U+0152-0153, U+02BB-02BC, U+02C6, U+02DA, U+02DC,
            U+2000-206F, U+2074, U+20AC, U+2122, U+2191, U+2193, U+2212, U+2215, U+FEFF, U+FFFD;
    }
    @font-face {
        font-family: "Roboto Slab";
        font-style: normal;
        font-weight: 200;
        font-display: swap;
        src: url(https://fonts.gstatic.com/s/robotoslab/v10/BngbUXZYTXPIvIBgJJSb6s3BzlRRfKOFbvjoDISmb2RlV9Su1cai.woff)
            format("woff");
        unicode-range: U+0000-00FF, U+0131, U+0152-0153, U+02BB-02BC, U+02C6, U+02DA, U+02DC,
            U+2000-206F, U+2074, U+20AC, U+2122, U+2191, U+2193, U+2212, U+2215, U+FEFF, U+FFFD;
    }
    @font-face {
        font-family: "Roboto Slab";
        font-style: normal;
        font-weight: 600;
        font-display: swap;
        src: url(https://fonts.gstatic.com/s/robotoslab/v10/BngMUXZYTXPIvIBgJJSb6ufN5qWr4xCC.woff2)
            format("woff2");
        unicode-range: U+0000-00FF, U+0131, U+0152-0153, U+02BB-02BC, U+02C6, U+02DA, U+02DC,
            U+2000-206F, U+2074, U+20AC, U+2122, U+2191, U+2193, U+2212, U+2215, U+FEFF, U+FFFD;
    }

    @font-face {
        font-family: "Open Sans";
        font-style: normal;
        font-weight: 400;
        font-display: swap;
        src: local("Open Sans Regular"), local("OpenSans-Regular"),
            url(https://fonts.gstatic.com/s/opensans/v17/mem8YaGs126MiZpBA-UFVZ0bf8pkAg.woff2)
                format("woff2");
        unicode-range: U+0000-00FF, U+0131, U+0152-0153, U+02BB-02BC, U+02C6, U+02DA, U+02DC,
            U+2000-206F, U+2074, U+20AC, U+2122, U+2191, U+2193, U+2212, U+2215, U+FEFF, U+FFFD;
    }
`;

export const App: React.FC = () => {
    const [appState, setAppState] = useState<IAppState>({
        inGameMode: false,
        guesses: [],
        remaining: [],
        packages: []
    });

    const context: IAppContext = {
        appState,
        setAppState
    };

    return (
        <AppContext.Provider value={context}>
            <BrowserRouter>
                <Global styles={[globalStyle, globalFocusStyle]} />
                <Header />
                <Content>
                    <Switch>
                        <Route path="/package">
                            <Suspense fallback={<LoadingIndicator />}>
                                <Package />
                            </Suspense>
                        </Route>
                        <Route path="/results">
                            <Suspense fallback={<LoadingIndicator />}>
                                <Results />
                            </Suspense>
                        </Route>
                        <Route path="/overview">
                            <Suspense fallback={<LoadingIndicator />}>
                                <Overview />
                            </Suspense>
                        </Route>
                        <Route path="/">
                            <Suspense fallback={<LoadingIndicator />}>
                                <Index />
                            </Suspense>
                        </Route>
                        <Redirect to="/" />
                    </Switch>
                </Content>
                <Footer />
            </BrowserRouter>
        </AppContext.Provider>
    );
};
