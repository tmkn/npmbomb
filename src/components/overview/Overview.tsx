/** @jsx jsx */
import { jsx, css, keyframes } from "@emotion/core";
import React, { useContext, useState, useEffect, useRef, memo } from "react";
import { BehaviorSubject } from "rxjs";
import { debounceTime, distinctUntilChanged, filter } from "rxjs/operators";
import { List, ListRowRenderer, AutoSizer } from "react-virtualized";

import {
    mq,
    serifFont,
    sansSerifFont,
    primaryColor,
    secondaryColor,
    textColor,
    secondaryColorLight,
    primaryColorLight
} from "../../css";
import { useAvailablePackagesLoader } from "../index/Index";
import { AppContext } from "../../AppContext";
import { Info } from "../shared/info/Info";
import { LoadingIndicator } from "../shared/loading/LoadingIndicator";
import { TokenHighlighter, TokenFormatter } from "../tokenhighlighter/TokenHighlighter";

interface IOverviewContext {
    query: string;
    setQuery: (quer: string) => void;
}

const OverviewContext = React.createContext<IOverviewContext>({
    query: "",
    setQuery() {}
});

const overviewStyle = css({
    [mq[0]]: {
        a: {
            display: "flex",
            paddingBottom: "0.5rem"
        },
        input: {
            flex: 1,
            color: primaryColor,
            border: `1px solid ${primaryColorLight}`,
            padding: "0.5rem",
            boxSizing: "border-box",
            outline: "none"
        }
    }
});

function matches(name: string, query: string): boolean {
    return new RegExp(`${query}`).test(name);
}

export default () => {
    const { loading, error } = useAvailablePackagesLoader();
    const {
        appState: { packages }
    } = useContext(AppContext);
    const [query, setQuery] = useState<string>("");
    const context: IOverviewContext = {
        query: query,
        setQuery: setQuery
    };

    if (error) return <Info>Couldn't load info</Info>;

    if (loading) return <LoadingIndicator />;

    const filteredPackages = packages.filter(p => matches(p, query)).sort();

    return (
        <OverviewContext.Provider value={context}>
            <div css={overviewStyle}>
                <h1>Overview</h1>
                <OverviewSearch />
                <Info>
                    {filteredPackages.length === 0 ? (
                        <NoOverviewResults />
                    ) : (
                        <ResultListing packages={filteredPackages} />
                    )}
                </Info>
            </div>
        </OverviewContext.Provider>
    );
};

const ResultListing: React.FC<{ packages: string[] }> = ({ packages }) => {
    const countStyle = css({
        [mq[0]]: {
            fontFamily: serifFont,
            marginBottom: "1rem"
        }
    });

    const resultsStyle = css({
        [mq[0]]: {
            minHeight: `300px`
        }
    });

    const rowStyle = css({
        overflow: "auto"
    });

    const rowRenderer: ListRowRenderer = ({ key, index, isScrolling, isVisible, style }) => {
        const p = packages[index];

        return (
            <div key={key} style={style} css={rowStyle}>
                <OverviewLink key={p} p={p} />
            </div>
        );
    };

    return (
        <React.Fragment>
            <span css={countStyle}>Found {packages.length} packages</span>
            <div css={resultsStyle}>
                <AutoSizer defaultHeight={600}>
                    {({ height, width }) => (
                        <List
                            overscanRowCount={10}
                            width={width}
                            height={height}
                            rowCount={packages.length}
                            rowHeight={30}
                            rowRenderer={rowRenderer}
                        />
                    )}
                </AutoSizer>
            </div>
        </React.Fragment>
    );
};

const tokenFormatterStyle = css({
    [mq[0]]: {
        fontWeight: "bold"
    }
});

const tokenFormatter: TokenFormatter = (token, i) => {
    return (
        <span key={i} css={tokenFormatterStyle}>
            {token}
        </span>
    );
};

const overviewLinkStyle = css({
    [mq[0]]: {
        textDecoration: "underline",
        textDecorationColor: secondaryColor,
        color: primaryColor,
        cursor: "pointer",
        whiteSpace: "pre",
        paddingLeft: "1rem",
        ":hover": {
            backgroundColor: primaryColorLight,
            color: secondaryColorLight
        }
    }
});

const OverviewLink: React.FC<{ p: string }> = ({ p }) => {
    const { query } = useContext(OverviewContext);

    return (
        <a css={overviewLinkStyle} href={`/package/${p}`}>
            <TokenHighlighter text={p} highlight={query} formatter={tokenFormatter} />
        </a>
    );
};

const OverviewSearch: React.FC = () => {
    const inputEl = useRef<HTMLInputElement>(null);
    let [inputObserver] = useState(() => new BehaviorSubject<string>(null!));
    const [query, setQuery] = useState<string>("");
    const overviewContext = useContext(OverviewContext);

    useEffect(() => {
        inputObserver
            .pipe(debounceTime(250))
            .pipe(filter(value => value !== null))
            .pipe(distinctUntilChanged())
            .subscribe(value => {
                overviewContext.setQuery(value);
            });

        return () => {
            inputObserver.unsubscribe();
        };
    }, []);

    function doQuery(e: React.ChangeEvent<HTMLInputElement>) {
        inputObserver.next(e.target.value);
        setQuery(e.target.value);
    }

    return (
        <Info>
            <input
                ref={inputEl}
                type="text"
                placeholder={`Search packages`}
                value={query}
                onChange={doQuery}
            />
        </Info>
    );
};

const NoOverviewResults: React.FC = () => {
    const { query } = useContext(OverviewContext);

    return <React.Fragment>No packages found that contain "{query}"</React.Fragment>;
};
