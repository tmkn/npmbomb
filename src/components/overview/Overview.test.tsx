import React from "react";

import { render, fireEvent, waitForElement, wait } from "@testing-library/react";

import Overview from "./Overview";
import { AppContext, IAppContext } from "../../App";

describe("<Overview />", () => {
    const packageNames = ["test", "foo-test", "@foo/test", "@blabla/foo"];
    const mockedAppContext: IAppContext = {
        appState: {
            guesses: [],
            inGameMode: false,
            remaining: [],
            packages: packageNames
        },
        setAppState: () => {}
    };

    beforeEach(() => {
        fetchMock.resetMocks();

        //needed by react-virtualized for rendering the list
        jest.spyOn(HTMLElement.prototype, "offsetHeight", "get").mockReturnValue(300);
        jest.spyOn(HTMLElement.prototype, "offsetWidth", "get").mockReturnValue(300);
    });

    test("matches snapshot", async () => {
        const { asFragment } = render(
            <AppContext.Provider value={mockedAppContext}>
                <Overview />
            </AppContext.Provider>
        );

        expect(asFragment()).toMatchSnapshot();

        await wait();
    });

    test("correctly displays packages", async () => {
        const { getByText, container } = render(
            <AppContext.Provider value={mockedAppContext}>
                <Overview />
            </AppContext.Provider>
        );

        for (const name of packageNames) {
            const element = (await waitForElement(() => getByText(name))) as HTMLAnchorElement;

            expect(element.tagName).toBe("A");
        }

        expect(container.getElementsByTagName("a").length).toEqual(packageNames.length);
    });

    test("correctly displays packages after search", async () => {
        const { getByText, container } = render(
            <AppContext.Provider value={mockedAppContext}>
                <Overview />
            </AppContext.Provider>
        );
        await wait();

        const inputEl = await waitForElement(() => container.querySelector("input[type='text']"));

        fireEvent.change(inputEl!, { target: { value: "test" } });

        await waitForElement(() => getByText(/Found 3 packages/), { timeout: 1000 });

        const filtered = await waitForElement(() => container.getElementsByTagName("a"));

        expect(filtered.length).toEqual(3);
    });

    test("correctly displays empty search results", async () => {
        const { getByText, container } = render(
            <AppContext.Provider value={mockedAppContext}>
                <Overview />
            </AppContext.Provider>
        );
        await wait();

        const inputEl = await waitForElement(() => container.querySelector("input[type='text']"));

        fireEvent.change(inputEl!, { target: { value: "adfasfsdf" } });

        await waitForElement(() => getByText(/No packages found/));

        const filtered = await waitForElement(() => container.getElementsByTagName("a"));

        expect(filtered.length).toEqual(0);
    });

    test("correctly displays loading error", async () => {
        fetchMock.mockReject();

        const { getByText, container } = render(
            <AppContext.Provider value={mockedAppContext}>
                <Overview />
            </AppContext.Provider>
        );

        await waitForElement(() => getByText(/Couldn't load info/));
    });
});
