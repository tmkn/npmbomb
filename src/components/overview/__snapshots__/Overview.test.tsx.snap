// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Overview /> matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-3f2zt9-Center"
  >
    <div
      class="css-1k3y3xi"
    >
      <div />
      <div />
    </div>
  </div>
</DocumentFragment>
`;
